cpp/ql/lib/semmle/code/cpp/models/interfaces/FormattingFunction.qll

```diff
@@ -125,6 +125,7 @@ abstract class FormattingFunction extends ArrayFunction, TaintFunction {
     // The formatting function either has a definition in the snapshot, or all
     // `DeclarationEntry`s agree on the number of parameters (otherwise we don't
     // really know the correct number)
+    result > 0 and // Avoid invalid declarations
```

Comment:
It would make more sense to me to make this condition part of the charPred of this abstract class. Reasoning: there are no arguments besides vararg ones, then there is no explicit argument for the formatting string, which would be very odd for a formatting function.

Maybe something like:
```ql
abstract class FormattingFunction extends ArrayFunction, TaintFunction {
  int firstFormatArgumentIndex;
  int numberOfExplicitParameters;

  FormattingFunction() {
    forex(FunctionDeclarationEntry fde | fde = this.getAnExplicitDeclarationEntry() |
      numberOfExplicitParameters = fde.getNumberOfParameters()
    ) and
    (
      if this.hasDefinition()
      then firstFormatArgumentIndex = this.getDefinition().getNumberOfParameters()
      else firstFormatArgumentIndex = numberOfExplicitParameters
    ) and
    firstFormatArgumentIndex > 0
  }
```
and implement the relevant `get` predicates in terms of `firstFormatArgumentIndex` and `numberOfExplicitParameters`.

I think it would also make sense at this point to move `getAnExplicitDeclarationEntry` to `Function`.

Reply:
Will this also work for intrinsic functions like `__builtin___sprintf_chk`? I will try.

Reply:
Intrinsic functions should still have a prototype in the database.

Reply:
I tried this out, and encountered a couple of problems:

1. Certain dataflow functions no longer identify the formatting functions as sinks if the functions have not been declared. Maybe this is ok, but I needed to fix up some of the qltests to add missing function declarations.
2. The builtin functions have no declaration entries - I tested this using `count(fn.getADeclarationEntry())` which returned `0`. This is fine, but I did need to add a special case to `FormattingFunction`.

I'll need to run DCA on this (once the tests pass), fix changenotes, and fix the tests on semmle-code.

Are you still ok with this design? @jketema

Reply:
> The builtin functions have no declaration entries - I tested this using `count(fn.getADeclarationEntry())` which returned `0`. This is fine, but I did need to add a special case to `FormattingFunction`.

That's an extractor bug.

Reply:
Using the `sprintf` macro invokes the builtin at `file://:0:0:0:0`, not the one at `builtins.cpp:4:5:4:27`. I guess I will dig into the TRAP files to see what happened.

Reply:
The duplication I expect, because there's both a builtin function, and one you define explicitly on your source file. What I don't expect is the declaration entry missing for the actual builtin.

Reply:
Fixed!

```c++
typedef unsigned long size_t;

extern "C" int __builtin___sprintf_chk(char *s, int flag, size_t os, const char *fmt, ...);
extern "C" int __builtin___snprintf_chk(char *s, size_t maxlen, int flag, size_t os,
                             const char *fmt, ...);
```

Reply:
So where is`#114=@"fun_decl___builtin___sprintf_chk(?)"` coming from, it looks like a function with just var args.

Reply:
My guess is that the call to `__builtin___sprint_chk` ignores the C++ declaration and just calls an implicit C function which appears in `#114`.  The declarations of the builtin functions appear to be optional, in that the code compiles perfectly fine without them.

Reply:
Correct, because our frontend defines those builtins. It just doesn't explain why `#114` looks so "funny".

Reply:
I think the two declarations were always there, but call was made to `#114`, ignoring `#15f`. The call to `#114` was treated as a valid call, but with the change, `#114` was no longer considered a valid `FormattingFunction` call target so the results disappeared and the test started to fail.

`extern "C"` makes the two labels identical.

Reply:
Ok, I'd like to better understand what is going on here before we merge this, because I'd like to be sure this isn't going to cause any problems on customer code. I'll spend some time on this tomorrow.

Reply:
We look at declaration entries in order to figure out if the declaration was implicit or not. There can be spurious function parameters caused by extraction errors, so we need to double-check the function declaration entries.

Reply:
I've opened a PR for an additional internal test related to this. I *think* the right solution here is to use the information from the `Function` and not the `FunctionDeclarationEntry` when we see that the function is a builtin function. In that case the information from the `Function` should be reliable, because the function's signature is under control of out frontend.

Reply:
I originally added `if this instanceof BuiltinFunction then firstFormatArgumentIndex = this.getNumberOfParameters()`, because `getNumberOfParameters()` was reliable for builtins. I'm happy to add this back, because we shouldn't rely on `FunctionDeclarationEntry` for builtins. If it's ok with you, shall this be the final solution?

Reply:
That sounds good.